### PR TITLE
[TEX-537] Optional support for unframing `grpc-web+json` requests.

### PR DESCRIPTION
**Pull request by timdawborn**
Add a new option for unframing `application/grpc-web+json` requests to be just `application/json` requests if the request is to a whitelisted hostname. This unframing is performed by mutating the received request before doing anything else with it, including passing it upstream or searching for existing tapes. This has the effect that, for all intents and purposes, the request is an `application/json` request for the purposes of persistence, similarity scoring, and proxying.

### CODE REVIEW

File: src/tests/passthrough.spec.ts [hunk 1]
```diff
@@ -36,3 +37,85 @@ describe("Passthrough", () => {
...
+      responseType: "arraybuffer",
+    });
+    expect(response.data).toEqual(BINARY_RESPONSE);
+  });
+
+  test("can pick any tape", async () => {
+    await axios.post(`${PROXAY_HOST}/__proxay/tape`, {
+      tape: "new-tape",
+    });
+  });
```
Comment (timdawborn) [inline src/tests/passthrough.spec.ts:65]: These are all copies of the above tests, but with the proxay server running with unframing enabled. I duplicated these to ensure they weren't affected by running with unframing enabled.

File: src/tests/passthrough.spec.ts [hunk 2]
```diff
@@ -36,3 +37,85 @@ describe("Passthrough", () => {
...
+          "content-type": "application/grpc-web+json",
+          host: `localhost:${TEST_SERVER_PORT}`,
+        },
+      }
+    );
+    expect(response.headers["content-type"]).toBe(
+      "application/json; charset=utf-8"
+    );
+    expect(response.data).toEqual({ email: "foo.bar@example.com" });
+  });
```
Comment (timdawborn) [inline src/tests/passthrough.spec.ts:120]: This is a new test for unframing.

File: src/server.ts [hunk 3]
```diff
@@ -74,29 +78,27 @@ export class RecordReplayServer {
...
       }
 
-      if (
-        this.preventConditionalRequests &&
-        (req.method === "GET" || req.method === "HEAD")
-      ) {
-        // Headers are always coming in as lowercase.
-        delete req.headers["if-modified-since"];
-        delete req.headers["if-none-match"];
-      }
```
Comment (timdawborn) [inline src/server.ts:84]: Move this logic into the new `rewriteRequest` helper method that includes other potential request modifications.

File: src/server.ts [hunk 4]
```diff
@@ -240,6 +242,75 @@ export class RecordReplayServer {
...
+
+    // Potentially prevent 304 responses from being able to be generated.
+    if (
+      this.preventConditionalRequests &&
+      (request.method === "GET" || request.method === "HEAD")
+    ) {
+      // Headers are always coming in as lowercase.
+      delete request.headers["if-modified-since"];
+      delete request.headers["if-none-match"];
+    }
```
Comment (timdawborn) [inline src/server.ts:260]: This is moved logic.

File: src/server.ts [hunk 5]
```diff
@@ -240,6 +242,75 @@ export class RecordReplayServer {
...
+
+    // Potentially unframe a grpc-web+json request.
+    if (
+      request.method === "POST" &&
+      request.headers["content-type"] === "application/grpc-web+json" &&
+      hostname != null &&
+      this.unframeGrpcWebJsonRequestsHostnames.includes(hostname)
+    ) {
+      this.rewriteGrpcWebJsonRequest(request);
+    }
```
Comment (timdawborn) [inline src/server.ts:270]: This is new logic.

File: src/cli.ts [hunk 6]
```diff
@@ -78,6 +78,10 @@ async function main(argv: string[]) {
       "--https-ca <filename.pem>",
       "Enable HTTPS server where the certificate was generated by this CA. Useful if you are using a self-signed certificate. Also requires --https-key and --https-cert."
     )
+    .option(
+      "--unframe-grpc-web-json-requests-hostname [hostname...]",
+      "Rewrite received requests whose content-type is application/grpc-web+json to be application/json, mutating the body of the request accordingly. This is useful if you want plain text tape records rather than binary data. The gRPC server needs to support receiving unframed requests for this option to be useful."
```
Comment (njday) [inline src/cli.ts:83]: **Nit:** We could also include the details about hostname matching in the description here

**Review by njday — APPROVED** [review]
LGTM 👍

**Review by flavray — APPROVED** [review]
Sheep It

**Pull request merged**